### PR TITLE
Feature/improve six select

### LINKED ISSUE #128: Improvement for six-select
Proposal to improve the `six-select` component.

image

### Multiselect
1. Group selected items visually.
2. Use checkboxes instead of check icons to indicate multiselect mode.
3. Add action to select all options

### Single Select
~4. Place the check icon on the left side of selected items.~

### General
5. Ensure the component maintains its height by not wrapping chosen options to multiple lines
6. Enhance the visual design of the filter by adding a search icon and changes spacing
7. Retain focus on the filter when moving the mouse.
8. Decrease spacing between items.
9. Reduce the dropdown's height.
10. ~~If the dropdown opens upwards, position the filter at the bottom.~~
11. Replace dropdown arrow with `x` to unselect all

### Current
image

### PR DESCRIPTION
**Pull request by ibirrer**
### ❓ Type of change

- [x] 👌 Enhancement (improving an existing functionality like performance)

### 📚 Description

Resolves  #128 

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [x] It's submitted to the `main` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [ ] I have updated the documentation accordingly.
- [ ] All tests are passing
- [ ] New/updated tests are included
- [ ] I have updated the "upcoming" section inside _docs/changelog.md_ explaining the changes I contributed

If adding a **new feature**, the PR's description includes:

- [x] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

### CODE REVIEW

File: libraries/ui-library/src/components/six-textarea/six-textarea.tsx [hunk 1]
```diff
@@ -104,9 +106,11 @@ export class SixTextarea {
   @Prop() autofocus = false;
 
   /** The textarea's spellcheck attribute. */
+  // eslint-disable-next-line @stencil-community/reserved-member-names
   @Prop() spellcheck = false;
 
   /** The textarea's inputmode attribute. */
+  // eslint-disable-next-line @stencil-community/reserved-member-names
```
Comment (Riatomerehentroweard) [inline libraries/ui-library/src/components/six-textarea/six-textarea.tsx:113]: Definitely not a fan of adding eslint-disable comments in so many places. Maybe it makes more sense to completely remove this rule alltogether then?

**Comment by davidemarcoli** [comment]
Before:
image

After:
image

I think it looks better, if there is no spacing between the first item and the start of the container. At the bottom it is the same.
There is now also a gap between the dropdown trigger and the item container. Is that intentional?

**Comment by davidemarcoli** [comment]
I know, that the move of the checkmark to the left was part of the ticket. Non the less, i think it isn't visually appealing that there is such a big gap on the left side of every item, if there isn't anything selected. 

image


Maybe indicate that an item is selected by changing the background color.

It looks like most component libraries are doing it like this:

Angular Material
image

Bootstrap
image

Prime NG
image

**Comment by davidemarcoli** [comment]
I think the removal of the arrow icon, when clearable is enabled, makes it difficult to recognize it still as a select instead of a text field.

Select
image
 
Text Input
image

It looks exactly the same!

**Comment by davidemarcoli** [comment]
The "default-value" attribute doesn't seem to do anything. What was it supposed to do? Maybe it was already broken?

**Comment by davidemarcoli** [comment]
The item label and the checkbox in the multi select don't seem to be aligned correctly.
image

**Comment by ibirrer** [comment]
Missing check icon in [six-header](https://six-group.github.io/six-webcomponents/components/six-header.html):

**New (check icon missing):**
image

**Before (Correct)**
image

**Comment by davidemarcoli** [comment]
In my eyes it would look better if the group item has to same height as the normal ones and the text is bold

Currently
image

(in my eyes) improved
image

**Comment by davidemarcoli** [comment]
Async Filtering does't seem to work anymore. 
image

**Comment by davidemarcoli** [comment]
six-select is noticeably laggy
https://github.com/six-group/six-webcomponents/assets/69892203/ef62e8af-f88a-4f6a-aea6-817bac049fb1

**Comment by davidemarcoli** [comment]
six-select seems to be laggy

https://github.com/six-group/six-webcomponents/assets/69892203/e29c04e8-136f-4c47-8ed6-52b7e75e9b3f

**Comment by davidemarcoli** [comment]
When you press the mouse down on an item, move the mouse outside of the container and then release it, the item is still "hovered"

https://github.com/six-group/six-webcomponents/assets/69892203/aa6aa1b8-ebde-4f2a-a480-d89318907d03

File: libraries/ui-library/package.json [hunk 2]
```diff
@@ -41,20 +41,17 @@
...
-    "@stencil-community/eslint-plugin": "^0.5.0",
+    "@stencil-community/eslint-plugin": "^0.7.1",
     "@stencil/angular-output-target": "^0.8.2",
     "@stencil/react-output-target": "^0.5.3",
     "@stencil/sass": "^3.0.7",
     "@stencil/vue-output-target": "^0.8.7",
     "@types/jest": "27.0.3",
-    "@typescript-eslint/eslint-plugin": "^5.59.7",
-    "@typescript-eslint/parser": "^5.59.6",
-    "eslint": "^8.41.0",
```
Comment (Riatomerehentroweard) [inline libraries/ui-library/package.json:52]: why was it necessary to remove eslint?

File: libraries/ui-library/src/components/six-avatar/six-avatar.scss [hunk 3]
```diff
@@ -28,7 +28,7 @@
   outline: none;
 
   &:hover {
-    box-shadow: var(--six-elevation-4dp);
+    box-shadow: var(--six-elevation-2dp);
```
Comment (Riatomerehentroweard) [inline libraries/ui-library/src/components/six-avatar/six-avatar.scss:31]: How is this change related to six-select? 🤔

File: libraries/ui-library/src/components/six-card/six-card.scss [hunk 4]
```diff
@@ -2,8 +2,7 @@
 
 :host {
   display: inline-block;
-  box-shadow: var(--six-elevation-8dp);
-  border-radius: 0;
-  padding: 1rem;
-  margin-bottom: 0.6rem;
+  box-shadow: var(--six-elevation-4dp);
+  border: 1px solid var(--six-color-web-rock-100);
+  padding: var(--six-spacing-medium);
```
Comment (davidemarcoli) [inline libraries/ui-library/src/components/six-card/six-card.scss:7]: How is this related to the six-select improvements?

**Comment by davidemarcoli** [comment]
Weird flickering in the input field when pressing the "Select all" button.

https://github.com/six-group/six-webcomponents/assets/69892203/8d1f9155-8847-42d1-9547-05c9abe0cc2e

Also. wouldn't it make more sense, if the button says "Deselect all" once pressed?

File: libraries/ui-library/src/components/six-textarea/six-textarea.tsx (hunk 1 above)
Comment (ibirrer) [reply libraries/ui-library/src/components/six-textarea/six-textarea.tsx]: Me neither. I think it's fine to turn off this specific rule. But I suggest we keep `@stencil-community/ban-default-true` for which I disabled some in the code too, as it is helpful to prevent more boolean properties that default to true.

File: libraries/ui-library/src/components/six-avatar/six-avatar.scss (hunk 3 above)
Comment (ibirrer) [reply libraries/ui-library/src/components/six-avatar/six-avatar.scss]: I tweaked the base shadow CSS variables slightly to achieve a cleaner appearance for the dropdown. This change also affected how other components looked, so I adjusted their shadows too for consistency. 

After looking at it again, I noticed we have 16 (!) different shadow styles in the base CSS, which is way too many. Most aren't even used, so I now reduced the number of shadow types to three (small, medium and large). See commit 0bb01908d7b0485a0e6ec5f03042eeef41cd2462.

File: libraries/ui-library/src/components/six-card/six-card.scss (hunk 4 above)
Comment (ibirrer) [reply libraries/ui-library/src/components/six-card/six-card.scss]: See https://github.com/six-group/six-webcomponents/pull/171#discussion_r1457531792

File: docs/changelog.md [hunk 5]
```diff
@@ -11,18 +11,32 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/).
...
+  separated values.
+- Moved check icons to the right in `six-select`.
+- Replaced check icons with checkboxes for `six-select` multiselect.
+- Display selected items on top of dropdown for `six-select` multiselect.
+- Reduced vertical spacing in `six-menu` items.
+- Improved visuals of `six-dropdonw` and added search icon to filter input.
+- Changed default debounce value to 0 for `six-dropdown` filter.
+- Either show clear or expand icon in `six-select`, but never both.
+- **Breaking**: Merged shadow css variables `--six-elevation-` into `--six-shadow-` and reduced
+  variants to small, medium and large.
```
Comment (Riatomerehentroweard) [inline docs/changelog.md:31]: I appreciate the hard work that has gone into this extensive PR, however to enhance the review efficiency and reduce the risk of breaking things, I'd suggest splitting such big PRs into smaller, more focused ones ☺

**Comment by ibirrer** [comment]
> Weird flickering in the input field when pressing the "Select all" button.

The number of additional items is now computed dynamically, which can cause some flickering. I think it would be possible to prevent the flickering by hiding added items, before computing the available width. However, since we don't have a lot of resources to work on this, I think we can accept the flickering problem for now.

> Also. wouldn't it make more sense, if the button says "Deselect all" once pressed?

Good idea. Changed with 6d0ac0f1a8a0f29146705aef05f9e66eec427d89

File: docs/changelog.md (hunk 5 above)
Comment (ibirrer) [reply docs/changelog.md:31]: I absolutely understand your concerns regarding the size of the PR. Unfortunately, it is extremely difficult to split this PR into smaller ones because the changes are distributed across multiple components, including select, dropdown, menu, input and menu-item. Even though it could be done now, after the changes have already been made, it would take a considerable amount of time to reorganize the PR.

That said, I have tried to group the changes into single commits, which should hopefully help a little. The last few commits are not grouped as they fix issues identified in earlier commits, and it is now a bit difficult to rebase them.

**Review by Riatomerehentroweard — DISMISSED** [review]

**Review by davidemarcoli — DISMISSED** [review]

**Review by Riatomerehentroweard — APPROVED** [review]

File: libraries/ui-library/src/components/six-select/six-select.tsx [hunk 6]
```diff
@@ -643,21 +704,41 @@ export class SixSelect {
             />
           </div>
 
+          {/* Selection container */}
+          <div
+            class={{
+              'selection-container': true,
+              'selection-container--border': this.selectionContainerItems.length > 0,
```
Comment (philippd) [inline libraries/ui-library/src/components/six-select/six-select.tsx:711]: There should not be a border when all items are selected:

Screenshot 2024-01-23 at 17 00 42

**Review by philippd — CHANGES_REQUESTED** [review]

**Review by philippd — DISMISSED** [review]

**Review by philippd — APPROVED** [review]
